### PR TITLE
Don't remove cast when converting method group to object

### LINKED ISSUE #58804: IDE0004 Unnecessary cast false positive
**Version Used**: 

main

**Steps to Reproduce**:

```csharp
class C
{
    static object M(object o)
    {
        return (object)o.ToString;
    }
}
```

**Expected Behavior**:

No unnecessary cast is reported.

**Actual Behavior**:

Unnecessary cast is reported. Removing the cast will produce this warning:

> warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?

Extracted from https://github.com/dotnet/roslyn/issues/50043#issuecomment-924209443

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #58804

@davidwengier @CyrusNajmabadi  for review.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs [hunk 1]
```diff
@@ -81,6 +81,11 @@ public static bool IsUnnecessaryCast(CastExpressionSyntax cast, SemanticModel se
             if (IsRemovableBitwiseEnumNegation(cast, semanticModel, cancellationToken))
                 return true;
 
+            // Special case for converting a method group to object. The compiler issues a warning if the cast is removed:
+            // warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?
+            if (semanticModel.GetOperation(cast.Expression, cancellationToken)?.Kind == OperationKind.MethodReference)
+                return false;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:87]: i think you shoudl check that the actual casted type is a non-delegate type.

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs [hunk 2]
```diff
@@ -12644,5 +12644,26 @@ void M(Guid g1, Guid g2)
...
+    }
+}
+";
+            await new VerifyCS.Test
+            {
+                TestCode = code,
+                FixedCode = code,
+                LanguageVersion = LanguageVersion.CSharp10,
+            }.RunAsync();
+        }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12667]: please add test where the cast *is* removed if the return type is `Delegate`, `MulticastDelegate`, or things like `Func<string>`.  thanks!

**Review by CyrusNajmabadi — CHANGES_REQUESTED** [review]
would like more tests to show expected behavior.

**Comment by Youssef1313** [comment]
I think there are already more tests covering the bug you mentioned (they're failing on CI currently).

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs [hunk 3]
```diff
@@ -81,6 +81,14 @@ public static bool IsUnnecessaryCast(CastExpressionSyntax cast, SemanticModel se
             if (IsRemovableBitwiseEnumNegation(cast, semanticModel, cancellationToken))
                 return true;
 
+            // Special case for converting a method group to object. The compiler issues a warning if the cast is removed:
+            // warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?
+            if (semanticModel.GetOperation(cast.Expression, cancellationToken)?.Kind == OperationKind.MethodReference &&
+                semanticModel.GetTypeInfo(cast.Type, cancellationToken).Type is { SpecialType: SpecialType.System_Object })
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:87]:
```suggestion
                semanticModel.GetTypeInfo(cast.Type, cancellationToken).Type?.SpecialType is SpecialType.System_Object)
```

(i'm not a fan of an entire property pattern aroudn a single property for the purpose of a null check.  i feel like `?.` is the idiomatic way to alerady do that.

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12667]: just to verify, we have tests for these three cases?

**Review by CyrusNajmabadi — APPROVED** [review]
Signing off, if such tests do exist.

**Comment by CyrusNajmabadi** [comment]
@Youssef1313 can you verify: https://github.com/dotnet/roslyn/pull/58920/files#r789113726 ?

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12667]: Not for all the three I think. I'll get to adding the tests soon.

Comment (Youssef1313) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12667]: https://github.com/dotnet/roslyn/blob/bd1a3814c5eaf3da54c45f80bf4b57062a8d6e9d/src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs#L1954-L1970

@CyrusNajmabadi Is this expected for this test?

Comment (Youssef1313) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12667]: @CyrusNajmabadi This scenario is now broken (regression by this PR):

```csharp
using System;

class C
{
    static Delegate M(object o)
    {
        return (Delegate)(object)o.ToString;
    }
}
```

Is this the scenario you were looking for?

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Can you take a look? Thanks!

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs [hunk 4]
```diff
@@ -81,6 +81,20 @@ public static bool IsUnnecessaryCast(CastExpressionSyntax cast, SemanticModel se
...
+            var castExpressionOperation = semanticModel.GetOperation(cast.Expression, cancellationToken);
+            if (castExpressionOperation?.Kind == OperationKind.MethodReference &&
+                semanticModel.GetTypeInfo(cast.Type, cancellationToken).Type?.SpecialType is SpecialType.System_Object)
+            {
+                // If we have a double cast, report as unnecessary, e.g:
+                // (object)(object)MethodGroup
+                // (Delegate)(object)MethodGroup
+                // If we have a single object cast, don't report as unnecessary e.g:
+                // (object)MethodGroup
+                return castExpressionOperation.Parent?.Parent?.Kind == OperationKind.Conversion;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs:95]: Please add test where it is cast to object, then down to T, where T is constrained to a delegate type.

**Comment by CyrusNajmabadi** [comment]
Test failures look legitimate.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Can you take another look? Thanks!

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
thanks!

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Can you restart the failing checks please?

**Comment by CyrusNajmabadi** [comment]
Done!

**Pull request merged**